src/benchmarklib/tpch/tpch_db_generator.cpp

```diff
@@ -0,0 +1,320 @@
+    {opossum::TpchTable::Customer, CUST}, {opossum::TpchTable::Orders, ORDER},   {opossum::TpchTable::LineItem, LINE},
+    {opossum::TpchTable::Nation, NATION}, {opossum::TpchTable::Region, REGION}};
+
+template <typename DSSType, typename MKRetType, typename... Args>
+DSSType _call_dbgen_mk(size_t idx, MKRetType (*mk_fn)(DSS_HUGE, DSSType* val, Args...), opossum::TpchTable table,
+                       Args... args) {
+  /**
+   * Preserve calling scheme (row_start(); mk...(); row_stop(); as in dbgen's gen_tbl())
+   */
+
+  const auto dbgen_table_id = tpch_table_to_dbgen_id.at(table);
+
+  row_start(dbgen_table_id);
+
+  DSSType value;
+  mk_fn(idx, &value, std::forward<Args>(args)...);
+
+  row_stop(dbgen_table_id);
+
+  return value;
+}
+
+float _convert_money(DSS_HUGE cents) {
+  const auto dollars = cents / 100;
+  cents %= 100;
+  return dollars + (static_cast<float>(cents)) / 100.0f;
+}
+
+void _free_and_null_permutations(distribution* d) {
+  free(d->permute);
+  d->permute = NULL;
+}
+
+/**
+ * Call this after using dbgen to avoid memory leaks
+ */
+void _dbgen_cleanup() {
+  _free_and_null_permutations(&nations);
+  _free_and_null_permutations(&regions);
+  _free_and_null_permutations(&o_priority_set);
+  _free_and_null_permutations(&l_instruct_set);
+  _free_and_null_permutations(&l_smode_set);
+  _free_and_null_permutations(&l_category_set);
+  _free_and_null_permutations(&l_rflag_set);
+  _free_and_null_permutations(&c_mseg_set);
+  _free_and_null_permutations(&colors);
+  _free_and_null_permutations(&p_types_set);
+  _free_and_null_permutations(&p_cntr_set);
+  _free_and_null_permutations(&articles);
+  _free_and_null_permutations(&nouns);
+  _free_and_null_permutations(&adjectives);
+  _free_and_null_permutations(&adverbs);
+  _free_and_null_permutations(&prepositions);
+  _free_and_null_permutations(&verbs);
+  _free_and_null_permutations(&terminators);
+  _free_and_null_permutations(&auxillaries);
+  _free_and_null_permutations(&np);
+  _free_and_null_permutations(&vp);
+  _free_and_null_permutations(&grammar);
+
+  if (asc_date != NULL) {
+    for (size_t idx = 0; idx < TOTDATE; ++idx) {
+      free(asc_date[idx]);
+    }
+    free(asc_date);
+  }
+  asc_date = NULL;
+}
+
+}  // namespace
+
+namespace opossum {
+
+std::unordered_map<TpchTable, std::string> tpch_table_names = {
+    {TpchTable::Part, "part"},         {TpchTable::PartSupp, "partsupp"}, {TpchTable::Supplier, "supplier"},
+    {TpchTable::Customer, "customer"}, {TpchTable::Orders, "order"},      {TpchTable::LineItem, "lineitem"},
+    {TpchTable::Nation, "nation"},     {TpchTable::Region, "region"}};
+
+TpchDbGenerator::TpchDbGenerator(float scale_factor, uint32_t chunk_size)
+    : _scale_factor(scale_factor), _chunk_size(chunk_size) {}
+
+std::unordered_map<TpchTable, std::shared_ptr<Table>> TpchDbGenerator::generate() {
+  TableBuilder customer_builder(_chunk_size, customer_column_types, customer_column_names);
```

Comment:
I found this really confusing to read. It looked to me like a function declaration. Could you use `{}` instead of `()` to make it clearer or put the Class name to the right?

```c++
TableBuilder order_builder{_chunk_size, order_column_types, order_column_names};

// or

auto order_builder = TableBuilder{_chunk_size, order_column_types, order_column_names};
```

Btw, did you know that you used a c++17 feature here? Until recently, the type of a template class couldn’t be inferred from it’s constructor arguments.

Reply:
"I found this really confusing to read. It looked to me like a function declaration" No, you didn't. You just push for that funky hipster curly braces thing.

"Btw, did you know that you used a c++17 feature here" Yep ;)

Reply:
> "I found this really confusing to read. It looked to me like a function declaration" No, you didn't. (If you did, where did you see the param types?) You just push for that funky hipster curly braces thing. ;)

Honestly I did! I didn’t know it even was a template class to begin with.

Reply:
I don't know who would put function declaration into function bodies (well, except for the guys who did dbgen....) but if the way we have initialized objects for years is confusing, I will change it...